src/python/pants/engine/selectors.py

```diff
@@ -4,26 +4,31 @@
 import ast
 from dataclasses import dataclass
 from textwrap import dedent
-from typing import Any, Generator, Iterable, Tuple, Type, cast
+from typing import Any, Generator, Generic, Iterable, Optional, Tuple, Type, TypeVar, cast
 
 from pants.util.meta import frozen_after_init
 from pants.util.objects import TypeConstraint
 
 
+_Product = TypeVar("_Product")
+
+
 @frozen_after_init
 @dataclass(unsafe_hash=True)
-class Get:
+class Get(Generic[_Product]):
   """Experimental synchronous generator API.
 
   May be called equivalently as either:
-    # verbose form: Get(product, subject_declared_type, subject)
-    # shorthand form: Get(product, subject_declared_type(<constructor args for subject>))
+    # verbose form: Get[product](subject_declared_type, subject)
+    # shorthand form: Get[product](subject_declared_type(<constructor args for subject>))
```

Comment:
What about changing this to better clarify what is a Type vs object?

```
# verbose form: Get[ProductType](SubjectType, subject)
# subject form: Get[ProductType](SubjectType(<constructor args for subject>))
```

Reply:
To bikeshed a little more, I think `ProductType` could make the reader think that it's an actual type that exists, not a template. `<ProductType>`?

Reply:
oh, because these are actually named after the fields of the `Get` dataclass! We could potentially make that relationship more clear, but I think it might actually be more clear to describe the arguments in terms of the fields of the dataclass, as opposed to some abstract `ProductType`. Does that make sense?